README.md
## WebSockets_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/WebSockets_Generic.svg?)](https://www.ardu-badge.com/WebSockets_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/WebSockets_Generic.svg)](https://github.com/khoih-prog/WebSockets_Generic/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/WebSockets_Generic/blob/master/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/WebSockets_Generic.svg)](http://github.com/khoih-prog/WebSockets_Generic/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Important notes about Portenta_H7 Ethernet](#Important-notes-about-Portenta_H7-Ethernet)
* [Why do we need this WebSockets_Generic library](#why-do-we-need-this-websockets_generic-library)
  * [Features](#features)
  * [Supported features of RFC6455](#supported-features-of-rfc6455)
  * [Important Notes](#important-notes)
  * [Currently supported Boards](#currently-supported-boards)
  * [Currently supported WiFi shields/modules](#currently-supported-wifi-shieldsmodules)
  * [Currently supported Ethernet shields/modules](#currently-supported-ethernet-shieldsmodules)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Packages' Patches](#packages-patches)
  * [1. For Adafruit nRF52840 and nRF52832 boards](#1-for-adafruit-nRF52840-and-nRF52832-boards)
  * [2. For Teensy boards](#2-for-teensy-boards)
  * [3. For Arduino SAM DUE boards](#3-for-arduino-sam-due-boards)
  * [4. For Arduino SAMD boards](#4-for-arduino-samd-boards)
      * [For core version v1.8.10+](#for-core-version-v1810)
      * [For core version v1.8.9-](#for-core-version-v189-)
  * [5. For Adafruit SAMD boards](#5-for-adafruit-samd-boards)
  * [6. For Seeeduino SAMD boards](#6-for-seeeduino-samd-boards)
  * [7. For STM32 boards](#7-for-stm32-boards) 
    * [7.1. For STM32 boards to use LAN8720](#71-for-stm32-boards-to-use-lan8720)
    * [7.2. For STM32 boards to use Serial1](#72-for-stm32-boards-to-use-serial1)
  * [8. For RP2040-based boards using Earle Philhower arduino-pico core](#8-for-rp2040-based-boards-using-earle-philhower-arduino-pico-core)
    * [8.1. To use BOARD_NAME](#81-to-use-board_name)
    * [8.2. To avoid compile error relating to microsecondsToClockCycles](#82-to-avoid-compile-error-relating-to-microsecondstoclockcycles)
  * [9. For Portenta_H7 boards using Arduino IDE in Linux](#9-for-portenta_h7-boards-using-arduino-ide-in-linux)
  * [10. For RTL8720DN boards using AmebaD core](#10-for-rtl8720dn-boards-using-amebad-core)
  * [11. For SAMD21 and SAMD51 boards using ArduinoCore-fab-sam core](#11-For-SAMD21-and-SAMD51-boards-using-ArduinoCore-fab-sam-core)
  * [12. For Seeeduino RP2040 boards](#12-For-Seeeduino-RP2040-boards)
  * [13. For Seeeduino nRF52840 boards](#13-For-Seeeduino-nRF52840-boards)
* [Libraries' Patches](#libraries-patches)
  * [1. For application requiring 2K+ HTML page](#1-for-application-requiring-2k-html-page)
  * [2. For Ethernet library](#2-for-ethernet-library)
  * [3. For EthernetLarge library](#3-for-ethernetlarge-library)
  * [4. For Etherne2 library](#4-for-ethernet2-library)
  * [5. For Ethernet3 library](#5-for-ethernet3-library)
  * [6. For UIPEthernet library](#6-for-uipethernet-library)
  * [7. For fixing ESP32 compile error](#7-for-fixing-esp32-compile-error)
  * [8. For fixing ESP8266 compile error](#8-for-fixing-esp8266-compile-error)
* [HOWTO Use analogRead() with ESP32 running WiFi and/or BlueTooth (BT/BLE)](#howto-use-analogread-with-esp32-running-wifi-andor-bluetooth-btble)
* [Important Notes](#important-notes-1)
  * [Limitations](#limitations)
  * [Limitations for Async](#limitations-for-async)
  * [Originally Supported Hardware](#originally-supported-hardware)
  * [wss / SSL](#wss--ssl)
  * [ESP Async TCP](#esp-async-tcp)
* [How to use](#how-to-use)
* [High Level Client API](#high-level-client-api)
* [Configure SIO_PING_INTERVAL, SIO_PONG_TIMEOUT and SIO_DISCONNECT_TIMEOUT_COUNT](#Configure-SIO_PING_INTERVAL-SIO_PONG_TIMEOUT-and-SIO_DISCONNECT_TIMEOUT_COUNT)
* [Examples](#examples)
  * [For Generic boards using W5x00 Ethernet shield](#for-generic-boards-using-w5x00-ethernet-shield)
  * [For Generic boards using WiFiNINA](#for-generic-boards-using-wifinina)
  * [For Generic boards using ENC28J60 Ethernet shield and EthernetENC library](#for-generic-boards-using-enc28j60-ethernet-shield-and-ethernetenc-library)
  * [For WiFiNINA](#for-wifinina)
  * [For WiFi101](#for-wifi101)
  * [For W5x00 Ethernet shield](#for-w5x00-ethernet-shield)
  * [For ENC28J60 Ethernet shield using UIPEthernet library](#for-enc28j60-ethernet-shield-using-uipethernet-library)
  * [For ESP32 board](#for-esp32-board)
  * [For ESP8266 board](#for-esp8266-board)
  * [For SeeedStudio WIO Terminal using Realtek RTL8720DN WiFi](#for-seeedstudio-wio-terminal-using-realtek-rtl8720dn-wifi) 
  * [For STM32 boards using built-in LAN8742A Ethernet](#for-stm32-boards-using-built-in-lan8742a-ethernet)
  * [For Teensy boards using ENC28J60 Ethernet shield and EthernetENC library](#for-teensy-boards-using-enc28j60-ethernet-shield-and-ethernetenc-library)
  * [For Teensy 4.1 boards using NativeEthernet](#for-teensy-41-boards-using-nativeethernet)
  * [For Teensy boards using W5x00 Ethernet shield](#for-teensy-boards-using-w5x00-ethernet-shield)
  * [For Teensy boards using WiFiNINA](#for-teensy-boards-using-wifinina)
  * [For Teensy 4.1 boards using QNEthernet](#for-teensy-41-boards-using-qnethernet)
  * [Nodejs Socket.IO Test Server to use with examples](#nodejs-socketio-test-server-to-use-with-examples)
  * [For WT32_ETH01 board](#for-wt32_eth01-board) **New**
  * [For Portenta_H7 board](#for-Portenta_H7-board). **New**
    * [Portenta_Ethernet](#Portenta_Ethernet).
    * [WiFi](#WiFi)
  * [For ESP8266 Async WebSocket](#For-ESP8266-Async-WebSocket) **New**
  * [For RP2040W boards using built-in WiFi](#For-RP2040W-boards-using-built-in-WiFi) **New**
* [Example Generic_WebSocketClient_WiFiNINA](#example-generic_websocketclient_wifinina)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [1. Generic_WebSocketClient_EthernetENC on NRF52840_FEATHER](#1-generic_websocketclient_ethernetenc-on-nrf52840_feather)
  * [2. Generic_Ethernet_Blinds on NRF52840_FEATHER with ENC28J60 using EthernetENC Library](#2-generic_ethernet_blinds-on-nrf52840_feather-with-enc28j60-using-ethernetenc-library)
  * [3. WebSocketClientSocketIO_W5500 on NRF52840_FEATHER with W5500 using Ethernet_Generic Library](#3-websocketclientsocketio_w5500-on-nrf52840_feather-with-w5500-using-Ethernet_Generic-library)
  * [4. Generic_WebSocketClientSocketIO_EthernetENC on NRF52840_FEATHER with ENC28J60 using EthernetENC Library](#4-generic_websocketclientsocketio_ethernetenc-on-nrf52840_feather-with-enc28j60-using-ethernetenc-library) 
  * [5. WIOTerminal_WebSocketClientSSL on SeeedStudio SAMD51 WIO_TERMINAL with Realtek RTL8720DN WiFi using Seeed_Arduino_rpcWiFi Library](#5-wioterminal_websocketclientssl-on-seeedstudio-samd51-wio_terminal-with-realtek-rtl8720dn-wifi-using-seeed_arduino_rpcwifi-library)
  * [6. Generic_WebSocketClientSSL_WiFiNINA on Arduino SAMD21 Nano-33-IoT with WiFiNINA using WiFiNINA_Generic Library](#6-generic_websocketclientssl_wifinina-on-arduino-samd21-nano-33-iot-with-wifinina-using-wifinina_generic-library)
  * [7. WebSocketClientSocketIO_NINA on Arduino SAMD21 Nano-33-IoT with WiFiNINA using WiFiNINA_Generic Library](#7-websocketclientsocketio_nina-on-arduino-samd21-nano-33-iot-with-wifinina-using-wifinina_generic-library)
    * [7.1 Client](#71-client)
    * [7.2 Server](#72-server)
  * [8. Generic_WebSocketClientSocketIO_W5500 on NRF52840_FEATHER with W5x00 using Ethernet_Generic Library](#8-generic_websocketclientsocketio_w5500-on-nrf52840_feather-with-w5x00-using-Ethernet_Generic-library)
    * [8.1 Client](#81-client)
    * [8.2 Server](#82-server)
  * [9. WebSocketClientSocketIO_W5500 on RASPBERRY_PI_PICO with W5x00 using Ethernet_Generic Library](#9-websocketclientsocketio_w5500-on-raspberry_pi_pico-with-w5x00-using-Ethernet_Generic-library)
    * [9.1 Client](#91-client)
    * [9.2 Server](#92-server)
  * [10. WebSocketClientSocketIO_W5500 on MBED RASPBERRY_PI_PICO with W5x00 using Ethernet_Generic Library](#10-websocketclientsocketio_w5500-on-mbed-raspberry_pi_pico-with-w5x00-using-Ethernet_Generic-library)
    * [10.1 Client](#101-client)
    * [10.2 Server](#102-server)
  * [11. WT32_ETH01_WebSocketClientSocketIO on WT32_ETH01](#11-wt32_eth01_websocketclientsocketio-on-wt32_eth01)
    * [11.1 Client](#111-client)
    * [11.2 Server](#112-server)
  * [12. Teensy_WebSocketClientSocketIO_QNEthernet on TEENSY 4.1 using QNEthernet](#12-teensy_websocketclientsocketio_qnethernet-on-teensy-41-using-qnethernet)
    * [12.1 Client](#121-client)
    * [12.2 Server](#122-server)
  * [13. Portenta_H7_WebSocketClientSocketIO on PORTENTA_H7_M7 with Ethernet](#13-Portenta_H7_WebSocketClientSocketIO-on-PORTENTA_H7_M7-with-Ethernet)
    * [13.1 Client](#131-client)
    * [13.2 Server](#132-server)
  * [14. WebSocketClientSocketIO_WiFi on Portenta_H7 using WiFi](#14-WebSocketClientSocketIO_WiFi-on-Portenta_H7-using-WiFi)
    * [14.1 Client](#141-client)
    * [14.2 Server](#142-server)
  * [15. Portenta_H7_WebSocketClient_Sticky_SocketIO on PORTENTA_H7_M7 using Ethernet](#15-Portenta_H7_WebSocketClient_Sticky_SocketIO-on-PORTENTA_H7_M7-using-Ethernet)
  * [16. ESP32_WebSocketClientSocketIO on ESP32S3_DEV](#16-ESP32_WebSocketClientSocketIO-on-ESP32S3_DEV)
  * [17. ESP8266_AsyncWSClient on ESP8266_NODEMCU_ESP12E](#17-ESP8266_AsyncWSClient-on-ESP8266_NODEMCU_ESP12E)
  * [18. WebSocketClientSocketIO_RP2040W on RP2040W using WiFi](#18-WebSocketClientSocketIO_RP2040W-on-RP2040W-using-WiFi) **New**
  * [19. Generic_WebSocketClientSocketIO_WiFiNINA on Nano_RP2040_Connect using WiFiNINA](#19-Generic_WebSocketClientSocketIO_WiFiNINA-on-Nano_RP2040_Connect-using-WiFiNINA) **New**
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License and credits](#license-and-credits)
* [Copyright](#copyright)

---
---

### Important notes about Portenta_H7 Ethernet

**Portenta_H7 Ethernet** doesn't work in mbed_portenta core v3.0.0, and has been fixed in core v3.0.1. 
Check [[Portenta_H7] Ethernet not working in core v3.0.0 #436](https://github.com/arduino/ArduinoCore-mbed/issues/436).

Please use mbed_portenta core **v2.8.0- or v3.0.1+ for Portenta_H7 Ethernet**.

---
---

### Why do we need this [WebSockets_Generic library](https://github.com/khoih-prog/WebSockets_Generic)

#### Features

Many Web services require WebSockets library, which is so far written only for ESP8266/ESP32 boards. The ESP boards rely on this [Markus Sattler's **WebSockets Library**](https://github.com/Links2004/arduinoWebSockets) to connect to Alexa via Sinric or SinricPro skills.

This [WebSockets_Generic library](https://github.com/khoih-prog/WebSockets_Generic) is a RFC6455-based WebSocket Server and Client for Arduino boards. It is based on and modified from [Markus Sattler's **WebSockets Library**](https://github.com/Links2004/arduinoWebSockets) to provide support to many more boards, such as **WT32_ETH01 (ESP32 + LAN8720), Arduino SAMD21, Adafruit SAMD21/SAMD51, nRF52, STM32F/L/H/G/WB/MP1, Teensy, SAM DUE, RP2040-based boards, etc.** and enable those boards to use WebSockets services, including voice-control Alexa along with Blynk. Those supported boards can also run **WebSockets Server.** The WebSockets can be used with **ESP’s WiFi, WiFiNINA, WiFi101, W5x00, ENC28J60 Ethernet and Teensy 4.1 NativeEthernet/QNEthernet.**


#### Supported features of RFC6455

 - text frame
 - binary frame
 - connection close
 - ping
 - pong
 - continuation frame

#### Important Notes

1. Currently, the WebSocketServer feature is usable only for ESP8266/ESP32 (ESP32, ESP32_S2, ESP32_S3 and ESP32_C3) and WT32_ETH01 (ESP32 + LAN8720).
2. The timeline when to fix and reintroduce the WebSocketServer feature to other boards is not determined yet.
3. Please use the new [**WebSockets2_Generic Library**](https://github.com/khoih-prog/WebSockets2_Generic) if WebSocketServer is necessary. See [Issue 2](https://github.com/khoih-prog/WebSockets_Generic/issues/2), [Issue 3](https://github.com/khoih-prog/WebSockets_Generic/issues/3) and [Issue 4](https://github.com/khoih-prog/WebSockets_Generic/issues/4)

---


#### Currently supported Boards

This [**WebSockets_Generic** library](https://github.com/khoih-prog/WebSockets_Generic) currently supports these following boards:

 1. **nRF52 boards**, such as **AdaFruit Feather nRF52832, nRF52840 Express, BlueFruit Sense, Itsy-Bitsy nRF52840 Express, Metro nRF52840 Express, NINA_B302_ublox, NINA_B112_ublox, etc.**
 
 2. **SAMD21**
  - Arduino SAMD21: ZERO, MKRs, NANO_33_IOT, etc.
  - Adafruit SAMD21 (M0): ItsyBitsy M0, Feather M0, Feather M0 Express, Metro M0 Express, Circuit Playground Express, Trinket M0, PIRkey, Hallowing M0, Crickit M0, etc.
  - Seeeduino:  LoRaWAN, Zero, Femto M0, XIAO M0, Wio GPS Board, etc.
  
 3. **SAMD51**
  - Adafruit SAMD51 (M4): Metro M4, Grand Central M4, ItsyBitsy M4, Feather M4 Express, Trellis M4, Metro M4 AirLift Lite, MONSTER M4SK Express, Hallowing M4, etc.
  - Seeeduino: Wio Terminal, Grove UI Wireless
  
 4. **SAM DUE** 
 5. **Teensy (4.1, 4.0, 3.6, 3.5, 3,2, 3.1, 3.0)**
 6. **STM32F/L/H/G/WB/MP1 boards (with 32+K Flash)**
  - Nucleo-144
  - Nucleo-64
  - Discovery
  - Generic STM32F0, STM32F1, STM32F2, STM32F3, STM32F4, STM32F7 (with 64+K Flash): x8 and up
  - STM32L0, STM32L1, STM32L4
  - STM32G0, STM32G4
  - STM32H7
  - STM32WB
  - STM32MP1
  - LoRa boards
  - 3-D printer boards
  - Generic Flight Controllers

changelog.md
## WebSockets_Generic Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/WebSockets_Generic.svg?)](https://www.ardu-badge.com/WebSockets_Generic)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/WebSockets_Generic.svg)](https://github.com/khoih-prog/WebSockets_Generic/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/WebSockets_Generic/blob/master/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/WebSockets_Generic.svg)](http://github.com/khoih-prog/WebSockets_Generic/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of Contents

* [Changelog](#changelog)
  * [Release v2.16.1](#release-v2161)
  * [Release v2.16.0](#release-v2160)
  * [Release v2.15.0](#release-v2150)
  * [Release v2.14.2](#release-v2142)
  * [Release v2.14.1](#release-v2141)
  * [Release v2.14.0](#release-v2140)
  * [Release v2.13.0](#release-v2130)
  * [Major Release v2.12.0](#major-release-v2120)
  * [Release v2.11.1](#release-v2111)
  * [Major Release v2.11.0](#major-release-v2110)
  * [Release v2.10.1](#release-v2101)
  * [Major Release v2.10.0](#major-release-v2100)
  * [Major Release v2.9.0](#major-release-v290)
  * [Major Release v2.8.0](#major-release-v280)
  * [Major Release v2.7.0](#major-release-v270)
  * [Major Release v2.6.0](#major-release-v260)
  * [Release v2.5.1](#release-v251)
  * [Major Release v2.5.0](#major-release-v250)
  * [Release v2.4.1](#release-v241)
  * [Major Release v2.4.0](#major-release-v240)
  * [Release v2.3.4](#release-v234)
  * [Release v2.3.3](#release-v233)
  * [Release v2.3.2](#release-v232)
  * [Release v2.3.1](#release-v231)
  * [Release v2.2.3](#release-v223)
  * [Release v2.2.2](#release-v222)
  * [Release v2.2.1](#release-v221)
  * [Release v2.1.3](#release-v213)
* [Important Notes](#important-notes)
  * [Originally Supported Hardware](#originally-supported-hardware)
  * [New support from v2.10.0](#new-support-from-v2100)
  * [New support from v2.9.0](#new-support-from-v290)
  * [New support from v2.8.0](#new-support-from-v280)
  * [New support from v2.7.0](#new-support-from-v270)
  * [New support from v2.5.0](#new-support-from-v250)
  * [New support from v2.4.0](#new-support-from-v240)
  * [New support from v2.3.4](#new-support-from-v234)
  * [New support from v2.3.2](#new-support-from-v232)
  * [New support from v2.3.1](#new-support-from-v231)
  * [New support from v2.2.3](#new-support-from-v223)
  * [New support from v2.2.2](#new-support-from-v222)
  * [New support from v2.1.3](#new-support-from-v213)
---
---

## Changelog

### Release v2.16.1

1. Using new [`WiFi101_Generic library`](https://github.com/khoih-prog/WiFi101_Generic) for sending larger data
2. Add astyle using `allman` style. Restyle the library
3. Update `Packages' Patches`

### Release v2.16.0

1. Add support to **RP2040-based boards, such as Nano_RP2040_Connect, RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040**, using WiFiNINA or Ethernet
2. Add `WS` and `BearSSL WSS` support to `RP2040W` using `CYW43439 WiFi` with `arduino-pico` core
3. Optional user-defined 
- `WEBSOCKETS_TCP_TIMEOUT`, default 5,000ms
- `EIO_HEARTBEAT_INTERVAL`, default 20,000ms
- `SIO_PING_INTERVAL`, default 60,000ms
- `SIO_PONG_TIMEOUT`, default 90,000ms
- `SIO_DISCONNECT_TIMEOUT_COUNT`, default 5
4. Update `Packages' Patches`

### Release v2.15.0

1. Use new [Ethernet_Generic library](https://github.com/khoih-prog/Ethernet_Generic) as default for W5x00.
2. Sync with `arduinoWebSockets v2.3.6`
3. Add examples and rewrite all the examples to support new features
4. Update `Packages' Patches`


### Release v2.14.2

1. Fix Async bug for ESP8266 when using NETWORK_ESP8266_ASYNC
2. Add example [ESP8266_AsyncWSClient](examples/Async/ESP8266_AsyncWSClient)
3. Update `platformio.ini` to temporarily use ESP32 core v2.0.0+ and use [ESPAsyncTCP](https://github.com/me-no-dev/ESPAsyncTCP) library for ESP8266 Async
4. Update `Packages' Patches`

### Release v2.14.1

1. Fix setInsecure() bug for WIO_Terminal.
2. Update `Packages' Patches` for `Seeeduino` core. Be sure to use the `Packages_Patches` of this library version v2.14.1+ to avoid compiler error.

### Release v2.14.0

1. Suppress unnecessary warnings.
2. Don't use `deprecated sha.h` for ESP32 core v1.0.6+. Check [warning sha.h is deprecated, use sha_parallel_engine.h #738](https://github.com/Links2004/arduinoWebSockets/issues/738)
3. Optimize code by passing by `reference` instead of `value`
4. Modify nearly all examples
5. Update `Packages' Patches`

### Release v2.13.0

1. Add support to `ESP32_S3`.
2. Add `sIOtype_PING` and `sIOtype_PING` SocketIO events. Check [Ping events propagated to main code socketIO events #18](https://github.com/khoih-prog/WebSockets_Generic/issues/18)
3. Add [ESP32_WebSocketClientSocketIO](examples/ESP32_WebSocketClientSocketIO) example
4. Update `Packages' Patches`

### Major Release v2.12.0

1. Add SSL support to ESP32-based WT32_ETH01 boards using LAN8720 Ethernet
2. Add WT32_ETH01-related [WT32_ETH01_SSL_SIO example](examples/WT32_ETH01/WT32_ETH01_SSL_SIO)
3. Update `Packages' Patches`


### Release v2.11.1


src/libsha1/libsha1.c
/* from valgrind tests */

/* ================ sha1.c ================ */
/*
  SHA-1 in C
  By Steve Reid <steve@edmweb.com>
  100% Public Domain

  Test Vectors (from FIPS PUB 180-1)
  "abc"
  A9993E36 4706816A BA3E2571 7850C26C 9CD0D89D
  "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
  84983E44 1C3BD26E BAAE4AA1 F95129E5 E54670F1
  A million repetitions of "a"
  34AA973C D4C4DAA4 F61EEB2B DBAD2731 6534016F
*/

/* #define LITTLE_ENDIAN * This should be #define'd already, if true. */
/* #define SHA1HANDSOFF * Copies data before messing with it. */

#if !defined(ESP8266) && !defined(ESP32)

#define SHA1HANDSOFF

#include <stdio.h>
#include <string.h>
#include <stdint.h>

#include "libsha1.h"

#define rol(value, bits) (((value) << (bits)) | ((value) >> (32 - (bits))))

/* blk0() and blk() perform the initial expand. */
/* I got the idea of expanding during the round function from SSLeay */
#if BYTE_ORDER == LITTLE_ENDIAN
#define blk0(i) (block->l[i] = (rol(block->l[i],24)&0xFF00FF00) \
                                 |(rol(block->l[i],8)&0x00FF00FF))
#elif BYTE_ORDER == BIG_ENDIAN
#define blk0(i) block->l[i]
#else
#error "Endianness not defined!"
#endif

#define blk(i) (block->l[i&15] = rol(block->l[(i+13)&15]^block->l[(i+8)&15] \
                                     ^block->l[(i+2)&15]^block->l[i&15],1))

/* (R0+R1), R2, R3, R4 are the different operations used in SHA1 */
#define R0(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk0(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R1(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R2(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0x6ED9EBA1+rol(v,5);w=rol(w,30);
#define R3(v,w,x,y,z,i) z+=(((w|x)&y)|(w&x))+blk(i)+0x8F1BBCDC+rol(v,5);w=rol(w,30);
#define R4(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0xCA62C1D6+rol(v,5);w=rol(w,30);


/* Hash a single 512-bit block. This is the core of the algorithm. */

void SHA1Transform(uint32_t state[5], const unsigned char buffer[64])
{
  uint32_t a, b, c, d, e;
  typedef union
  {
    unsigned char c[64];
    uint32_t l[16];
  } CHAR64LONG16;

#ifdef SHA1HANDSOFF
  CHAR64LONG16 block[1];  /* use array to appear as a pointer */
  memcpy(block, buffer, 64);
#else
  /* The following had better never be used because it causes the
     pointer-to-const buffer to be cast into a pointer to non-const.
     And the result is written through.  I threw a "const" in, hoping
     this will cause a diagnostic.
  */
  CHAR64LONG16* block = (const CHAR64LONG16*)buffer;
#endif

  /* Copy context->state[] to working vars */
  a = state[0];
  b = state[1];
  c = state[2];
  d = state[3];
  e = state[4];
  /* 4 rounds of 20 operations each. Loop unrolled. */
  R0(a, b, c, d, e, 0);
  R0(e, a, b, c, d, 1);
  R0(d, e, a, b, c, 2);
  R0(c, d, e, a, b, 3);
  R0(b, c, d, e, a, 4);
  R0(a, b, c, d, e, 5);
  R0(e, a, b, c, d, 6);
  R0(d, e, a, b, c, 7);
  R0(c, d, e, a, b, 8);
  R0(b, c, d, e, a, 9);
  R0(a, b, c, d, e, 10);
  R0(e, a, b, c, d, 11);
  R0(d, e, a, b, c, 12);
  R0(c, d, e, a, b, 13);
  R0(b, c, d, e, a, 14);
  R0(a, b, c, d, e, 15);
  R1(e, a, b, c, d, 16);
  R1(d, e, a, b, c, 17);
  R1(c, d, e, a, b, 18);
  R1(b, c, d, e, a, 19);
  R2(a, b, c, d, e, 20);
  R2(e, a, b, c, d, 21);
  R2(d, e, a, b, c, 22);
  R2(c, d, e, a, b, 23);
  R2(b, c, d, e, a, 24);
  R2(a, b, c, d, e, 25);
  R2(e, a, b, c, d, 26);
  R2(d, e, a, b, c, 27);
  R2(c, d, e, a, b, 28);
  R2(b, c, d, e, a, 29);
  R2(a, b, c, d, e, 30);
  R2(e, a, b, c, d, 31);
  R2(d, e, a, b, c, 32);
  R2(c, d, e, a, b, 33);
  R2(b, c, d, e, a, 34);
  R2(a, b, c, d, e, 35);
  R2(e, a, b, c, d, 36);
  R2(d, e, a, b, c, 37);
  R2(c, d, e, a, b, 38);
  R2(b, c, d, e, a, 39);
  R3(a, b, c, d, e, 40);
  R3(e, a, b, c, d, 41);
  R3(d, e, a, b, c, 42);
  R3(c, d, e, a, b, 43);
  R3(b, c, d, e, a, 44);
  R3(a, b, c, d, e, 45);
  R3(e, a, b, c, d, 46);
  R3(d, e, a, b, c, 47);
  R3(c, d, e, a, b, 48);
  R3(b, c, d, e, a, 49);
  R3(a, b, c, d, e, 50);
  R3(e, a, b, c, d, 51);
  R3(d, e, a, b, c, 52);
  R3(c, d, e, a, b, 53);
  R3(b, c, d, e, a, 54);
  R3(a, b, c, d, e, 55);
  R3(e, a, b, c, d, 56);
  R3(d, e, a, b, c, 57);
  R3(c, d, e, a, b, 58);
  R3(b, c, d, e, a, 59);
  R4(a, b, c, d, e, 60);
  R4(e, a, b, c, d, 61);
  R4(d, e, a, b, c, 62);
  R4(c, d, e, a, b, 63);
  R4(b, c, d, e, a, 64);
  R4(a, b, c, d, e, 65);
  R4(e, a, b, c, d, 66);
  R4(d, e, a, b, c, 67);
  R4(c, d, e, a, b, 68);
  R4(b, c, d, e, a, 69);
  R4(a, b, c, d, e, 70);
  R4(e, a, b, c, d, 71);
  R4(d, e, a, b, c, 72);
  R4(c, d, e, a, b, 73);
  R4(b, c, d, e, a, 74);
  R4(a, b, c, d, e, 75);
  R4(e, a, b, c, d, 76);
  R4(d, e, a, b, c, 77);
  R4(c, d, e, a, b, 78);
  R4(b, c, d, e, a, 79);

  /* Add the working vars back into context.state[] */
  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;
  state[4] += e;
  /* Wipe variables */
  a = b = c = d = e = 0;

#ifdef SHA1HANDSOFF
  memset(block, '\0', sizeof(block));
#endif
}


/* SHA1Init - Initialize new context */

void SHA1Init(SHA1_CTX* context)
{
  /* SHA1 initialization constants */
  context->state[0] = 0x67452301;
  context->state[1] = 0xEFCDAB89;
  context->state[2] = 0x98BADCFE;
  context->state[3] = 0x10325476;
  context->state[4] = 0xC3D2E1F0;
  context->count[0] = context->count[1] = 0;
}


/* Run your data through this. */

void SHA1Update(SHA1_CTX* context, const unsigned char* data, uint32_t len)
{
  uint32_t i, j;

  j = context->count[0];

  if ((context->count[0] += len << 3) < j)
    context->count[1]++;

  context->count[1] += (len >> 29);
  j = (j >> 3) & 63;

  if ((j + len) > 63)
  {
    memcpy(&context->buffer[j], data, (i = 64 - j));
    SHA1Transform(context->state, context->buffer);

    for ( ; i + 63 < len; i += 64)
    {
      SHA1Transform(context->state, &data[i]);
    }

    j = 0;
  }
  else
    i = 0;

  memcpy(&context->buffer[j], &data[i], len - i);
}


/* Add padding and return the message digest. */

void SHA1Final(unsigned char digest[20], SHA1_CTX* context)
{
  unsigned i;
  unsigned char finalcount[8];
  unsigned char c;

#if 0  /* untested "improvement" by DHR */
  /* Convert context->count to a sequence of bytes
     in finalcount.  Second element first, but
     big-endian order within element.
     But we do it all backwards.
  */
  unsigned char *fcp = &finalcount[8];

  for (i = 0; i < 2; i++)
  {
    uint32_t t = context->count[i];
    int j;

    for (j = 0; j < 4; t >>= 8, j++)
      * --fcp = (unsigned char) t;
  }

#else

  for (i = 0; i < 8; i++)
  {
    /* Endian independent */
    finalcount[i] = (unsigned char)
                    ((context->count[(i >= 4 ? 0 : 1)] >> ((3 - (i & 3)) * 8) ) & 255);
  }

#endif

  c = 0200;
  SHA1Update(context, &c, 1);

  while ((context->count[0] & 504) != 448)
  {
    c = 0000;
    SHA1Update(context, &c, 1);
  }

  SHA1Update(context, finalcount, 8);  /* Should cause a SHA1Transform() */

  for (i = 0; i < 20; i++)
  {
    digest[i] = (unsigned char)
                ((context->state[i >> 2] >> ((3 - (i & 3)) * 8) ) & 255);
  }

  /* Wipe variables */
  memset(context, '\0', sizeof(*context));
  memset(&finalcount, '\0', sizeof(finalcount));
}
/* ================ end of sha1.c ================ */


#endif
